### PR TITLE
Update the CellId class.

### PR DESCRIPTION
**Pull request by bangerth**
This basically just applies the style we use everywhere else to this class,
    i.e., make the documentation more uniform, make some arguments 'const' or
    a reference, move function definitions out of line, etc.

### CODE REVIEW

**Comment by davydden** [comment]
LGTM

**Comment by bangerth** [comment]
Follow-up to #2981.

**Comment by bangerth** [comment]
I've pushed a second commit that renames the poorly named `id` member to `child_indices`.

**Comment by tjhei** [comment]
Sigh. Thanks for fixing up this horrible mess I produced.

**Comment by bangerth** [comment]
I've pushed a third commit that documents what's happening inside operator<< / >> a bit better.

File: include/deal.II/grid/cell_id.h [hunk 1]
```diff
@@ -83,42 +80,68 @@ class CellId
...
   /**
-   * compare two CellIds
+   * Compare two CellIds for inequality.
    */
   bool operator!= (const CellId &other) const;
 
   /**
-   * compare two CellIds
+   * Compare two CellIds with regard to an ordering. The details of this
+   * ordering are unspecified except that the oepration provides a
```
Comment (masterleinad) [inline include/deal.II/grid/cell_id.h]: `oepration`->`operation`

**Pull request merged**